homeassistant/components/notify/pushbullet.py

```diff
@@ -24,27 +27,81 @@ def get_service(hass, config):
         return None
 
     try:
-        return PushBulletNotificationService(config[CONF_API_KEY])
-
+        pushbullet = PushBullet(config[CONF_API_KEY])
     except InvalidKeyError:
         _LOGGER.error(
             "Wrong API key supplied. "
             "Get it at https://www.pushbullet.com/account")
         return None
 
+    return PushBulletNotificationService(pushbullet)
+
 
 # pylint: disable=too-few-public-methods
 class PushBulletNotificationService(BaseNotificationService):
     """ Implements notification service for Pushbullet. """
 
-    def __init__(self, api_key):
-        from pushbullet import Pushbullet
+    def __init__(self, pb):
+        self.pushbullet = pb
+        self.pbtargets = {}
+        self.refresh()
 
-        self.pushbullet = Pushbullet(api_key)
+    def refresh(self):
+        '''
+        Refresh devices, contacts, channels, etc
 
-    def send_message(self, message="", **kwargs):
-        """ Send a message to a user. """
+        pbtargets stores all targets available from this pushbullet instance
+        into a dict. These are PB objects!. It sacrifices a bit of memory
+        for faster processing at send_message
+        '''
+        self.pushbullet.refresh()
+        self.pbtargets = {
+            'devices':
+                {tgt.nickname: tgt for tgt in self.pushbullet.devices},
+            'contacts':
+                {tgt.email: tgt for tgt in self.pushbullet.contacts},
+            'channels':
+                {tgt.channel_tag: tgt for tgt in self.pushbullet.channels},
+        }
 
+    def send_message(self, message=None, **kwargs):
+        """
+        Send a message to a specified target.
+        If no target specified, a 'normal' push will be sent to all devices
+        linked to the PB account.
+        """
+        targets = kwargs.get(ATTR_TARGET)
+        # Disabeling title
         title = kwargs.get(ATTR_TITLE)
+        title = None
+
+        if targets:
+            # Make list if not so
+            if not isinstance(targets, list):
+                targets = [targets]
+
+            # Main loop, Process all targets specified
+            for ttype, tname in [target.split('.', 1) for target in targets]:
+                if ttype == 'device' and not tname:
+                    # Allow for 'normal' push, combined with other targets
+                    self.pushbullet.push_note(title, message)
+                    _LOGGER.info('Sent notification to self')
+                    continue
+
+                # Attempt push_note on a dict value. Keys are types & target
+                # name. The pbtargets have all *actual* targets.
+                try:
```

Comment:
It's not quite sure why a refresh is done when a `PushError` happens and not when a `KeyError` happens.

I would imagine the code to be like this:
_(remove caching contacts from the init)_

```python
refreshed = False

if ttype not in self.pbtargets.keys():
  _LOGGER.error('invalid type specified: %s', ttype)
  continue

if tname not in self.pbtargets[ttype] and not refreshed:
  self.refresh()
  refreshed = True

if tname not in self.pbtargets[ttype]:
  _LOGGER.error('Unable to find target: %s.%s', ttype, tname)
  continue

try:
  self.pbtargets[ttype][tname].push_note(…)
  _LOGGER.info('Sent notification to %s.%s', ttype, tname)
except self.pushbullet.errors.PushError:
  _LOGGER.error('Notify failed to: %s.%s', ttype, tname)
```